src/base/static/components/place-detail/place-detail-editor.js

```diff
@@ -0,0 +1,219 @@
+import emitter from "../../utils/emitter";
+import classNames from "classnames";
+import {
+  Map as ImmutableMap,
+  OrderedMap as ImmutableOrderedMap,
+} from "immutable";
+import Spinner from "react-spinner";
+import "react-spinner/react-spinner.css";
+
+import FormField from "../form-fields/form-field";
+import WarningMessagesContainer from "../ui-elements/warning-messages-container";
+import { scrollTo } from "../../utils/scroll-helpers";
+import { extractEmbeddedImages } from "../../utils/embedded-images";
+const Util = require("../../js/utils.js");
+
+import { placeDetailEditor as messages } from "../../messages";
+import constants from "../../constants";
+
+import "./place-detail-editor.scss";
+
+class PlaceDetailEditor extends Component {
+  constructor(props) {
+    super(props);
+
+    let fields = ImmutableOrderedMap();
+    this.props.categoryConfig.fields
+      // NOTE: In the editor, we have to strip out the submit field here,
+      // otherwise, since we don't render it at all, it will always be invalid.
+      .filter(field => field.type !== constants.SUBMIT_FIELD_TYPENAME)
+      .forEach(field => {
+        fields = fields.set(
+          field.name,
+          ImmutableMap().set(
+            constants.FIELD_STATE_VALUE_KEY,
+            this.props.placeModel.get(field.name)
+          )
+        );
+      });
+
+    this.state = {
+      fields: fields,
+      formValidationErrors: new Set(),
+      showValidityStatus: false,
+    };
+  }
+
+  componentWillMount() {
+    emitter.addListener("place-model:update", () => {
+      // Validate the form.
+      const newValidationErrors = this.state.fields
+        .filter(value => !value.get(constants.FIELD_STATE_VALIDITY_KEY))
+        .reduce((newValidationErrors, invalidField) => {
+          return newValidationErrors.add(
+            invalidField.get(constants.FIELD_STATE_VALIDITY_MESSAGE_KEY)
+          );
+        }, new Set());
+
+      if (newValidationErrors.size === 0) {
+        this.props.onModelIO(constants.PLACE_MODEL_IO_START_ACTION);
+
+        const attrs = this.state.fields
+          .filter(state => state.get(constants.FIELD_STATE_VALUE_KEY) !== null)
+          .map(state => state.get(constants.FIELD_STATE_VALUE_KEY))
+          .toJS();
+
+        if (this.state.fields.get(constants.GEOMETRY_PROPERTY_NAME)) {
+          attrs.style = this.geometryStyle;
+        }
+
+        // Replace image data in rich text fields with placeholders built from each
+        // image's name.
+        // TODO: This logic is better suited for the FormField component,
+        // perhaps in an onSave hook.
+        this.props.categoryConfig.fields
+          .filter(
+            field => field.type === constants.RICH_TEXTAREA_FIELD_TYPENAME
+          )
+          .forEach(field => {
+            attrs[field.name] = extractEmbeddedImages(attrs[field.name]);
+          });
+
+        this.props.onPlaceModelSave(attrs, {
+          success: model => {
```

Comment:
These deeply nested callbacks within callbacks could be extracted into variables or methods.

Reply:
I wasn't going to bring this up yet, but since you mention it, I think we should have an api client abstraction for our backend. It could be a separate module, like `shareabouts-api-client` or `mapseed-api-client`, that has an `onPlaceModelSave` function which accepts the required data as arguments..

Reply:
I see-- do you think it makes sense to do this now, or tackle it when we factor out the Backbone models later?

Reply:
It makes most sense to me to just extract them into methods on this same class/object for now, and do the thing @modulitos mentioned later.

Reply:
I think either approach is good enough for now. I don't want to hold up this PR, so if @zmbc 's approach is more feasible, let's go with that. We can always submit a PR later to create the api client abstraction.